Upload improvements

dodget opened the pull request:
This PR resolves [ATGU-2716](https://jira.huit.harvard.edu/browse/ATGU-2716) and [ATGU-2717](https://jira.huit.harvard.edu/browse/ATGU-2717) by:

- When using the PDS to get user data, if the user's `loginName` is null, get the email from the first entry of the user's `emails` list
- If the email is provided, but the user's full name has not been, we now query the PDS using the provided email in an attempt to get the user's full name.

## Notes
- I didn't see a way to execute a bulk query of the PDS using emails to retrieve information. Those queries have to be executed individually per email as needed

arthurian commented on harvardsettings_pi.php:
```diff
@@ -382,6 +389,37 @@ public function get_pds_emails($url) {
+            CURLOPT_USERPWD,
+            $this->CI->config->item('pds_client_id') . ":" . $this->CI->config->item('pds_client_secret')
+        );
+        $output = json_decode(curl_exec($c));
```
Note that [curl_exec()](https://www.php.net/curl_exec) returns `FALSE` if there's a failure of some kind, which will output as `NULL` after being json decoded. Might be worth adding a check for that.

arthurian commented on harvardsettings_pi.php:
```diff
@@ -382,6 +389,37 @@ public function get_pds_emails($url) {
+            }  
+        }
+        $accessing_user = $this->CI->data['login']->email." (".$this->CI->data['login']->user_id.")";
+        log_message(
```
This is great for auditing PDS access.

arthurian commented on harvardsettings_pi.php:
```diff
@@ -369,9 +375,10 @@ public function get_pds_emails($url) {
             else {
                 $name = $person->names[0]->firstName . " " . $person->names[0]->lastName;
             }
+            $email = $person->loginName ? $person->loginName : $person->emails[0]->email;
```
If a person has multiple email addresses, is the first email always the "official" email address? My assumption is yes, but I'm not sure if that's really true. I believe there's an `officialEmailIndicator` boolean  for each email entry that indicates whether it's official or not. This only matters as far as Harvard Key authentication is concerned, since the email will need to match and I think Harvard Key sends the official one.

dodget replied:
Ah, good to know about the `officialEmailIndicator`. Thanks!

arthurian reviewed (approved):
Looks good @dodget. I just had one question about retrieving the email if there's no `loginName` and a suggestion on the curl request.

dodget commented:
I've made the following improvements:

- Added logging if the `json_decode` returns `Null`
- I look for the email with `officialEmailIndicator` flag set to true
- Refactored curling the PDS to a private function for an extra layer of isolation for the service
